### PR TITLE
zocl doc

### PR DESCRIPTION
**Pull request by chvamshi-xilinx**
Hi 

This change contains a documentation of all  zocl IOCTLS

Thanks

### CODE REVIEW

File: src/runtime_src/core/edge/include/zynq_ioctl.h [hunk 1]
```diff
@@ -27,21 +75,39 @@
...
 	DRM_ZOCL_EXECBUF,
+	/* Read the xclbin and map CUs. Used in non-PR platform */
 	DRM_ZOCL_READ_AXLF,
+	/* Get the soft kernel command */
 	DRM_ZOCL_SK_GETCMD,
+	/* Create the soft kernel */
 	DRM_ZOCL_SK_CREATE,
+	/* Report the soft kernel state */
 	DRM_ZOCL_SK_REPORT,
+	/* Get the physical address and apt index */
```
Comment (mamin506) [inline src/runtime_src/core/edge/include/zynq_ioctl.h]: Similar as DRM_ZOCL_INFO_BO?

File: src/runtime_src/core/edge/include/zynq_ioctl.h [hunk 2]
```diff
@@ -18,6 +18,54 @@
...
  */
 
+/**
+ * DOC: A GEM style driver for Xilinx edge based accelerators
+ * This file defines ioctl command codes and associated structures for interacting with
+ * *zocl* driver for Xilinx FPGA platforms(Zynq/ZynqMP/Versal).
+ *
+ * Device memory allocation is modeled as buffer objects (bo). For each bo driver tracks the host pointer
+ * backed by scatter gather list -- which provides backing storage on host -- and the corresponding device
+ * side allocation of contiguous buffer in one of the memory mapped RAM/DDRs/BRAMs, etc.
```
[the remark this answers is not in the document]
Comment (sonals) [reply src/runtime_src/core/edge/include/zynq_ioctl.h]: 4. Support for SMMU based virtual memory and CMA based contiguous physical buffers
5. PL DDRs are always accessed as CMA buffers. Both Linux and PL logic can access PL DDRs

File: src/runtime_src/core/edge/include/zynq_ioctl.h [hunk 3]
```diff
@@ -18,6 +18,54 @@
...
+ *      object
+ * 7    Update bo backing storage with user's  DRM_IOCTL_ZOCL_PWRITE_BO       drm_zocl_pwrite_bo
+ *      data
+ * 8    Read back data in bo backing storage   DRM_IOCTL_ZOCL_PREAD_BO        drm_zocl_pread_bo
+ * 9    Update device view with a specific     DRM_IOCTL_ZOCL_PCAP_DOWNLOAD   drm_zocl_pcap_download
+ *      xclbin image
+ * 10   Read the xclbin and map the compute    DRM_IOCTL_ZOCL_READ_AXLF       drm_zocl_axlf 
+ *      units. Used in non-PR platforms
+ * 11   Send an execute job to a compute unit  DRM_IOCTL_ZOCL_EXECBUF         drm_zocl_execbuf
+ * 12   Get the soft kernel command            DRM_IOCTL_ZOCL_SK_GETCMD       drm_zocl_sk_getcmd
```
Comment (sonals) [inline src/runtime_src/core/edge/include/zynq_ioctl.h]: Call out the next three (13, 14 and 15) as experimental ioctls. Will do security review before enabling it for general use.

File: src/runtime_src/core/edge/include/zynq_ioctl.h [hunk 4]
```diff
@@ -27,21 +75,39 @@
...
+	/* Buffer creation */
 	DRM_ZOCL_CREATE_BO = 0,
+	/* Buffer creation from user provided pointer */
 	DRM_ZOCL_USERPTR_BO,
+	/* Get the buffer handle of given physical address */
 	DRM_ZOCL_GET_HOST_BO,
+	/* Map buffer into application user space (no DMA is performed) */
 	DRM_ZOCL_MAP_BO,
+	/* Sync buffer (like fsync) in the desired direction by using DMA */
 	DRM_ZOCL_SYNC_BO,
```
Comment (sonals) [inline src/runtime_src/core/edge/include/zynq_ioctl.h]: cache flush/invalidation of cache able buffers.

File: src/runtime_src/core/edge/include/zynq_ioctl.h (hunk 3 above)
Comment (chvamshi-xilinx) [reply src/runtime_src/core/edge/include/zynq_ioctl.h]: @sonal, you mean to say 12 , 13 , 14 which are for soft kernels. 15 is about CU_INFO

File: src/runtime_src/core/edge/include/zynq_ioctl.h [hunk 5]
```diff
@@ -18,6 +18,54 @@
...
+ * Device memory allocation is modeled as buffer objects (bo). For each bo driver tracks the host pointer
+ * backed by scatter gather list -- which provides backing storage on host -- and the corresponding device
+ * side allocation of contiguous buffer in one of the memory mapped RAM/DDRs/BRAMs, etc.
+ *
+ * Exection model is asynchronous where execute commands are submitted using command buffers and POSIX poll
+ * is used to wait for finished commands. Commands for a compute unit can only be submitted after an explicit
+ * context has been opened by the client.
+ *
+ * *zocl* driver functionality is described in the following table. All the APIs are multi-threading and
+ * multi-process safe.
```
[the remark this answers is not in the document]
Comment (chvamshi-xilinx) [reply src/runtime_src/core/edge/include/zynq_ioctl.h]: @mamin506 , You meant to say **i dont think thos APIs are multi threaded safe**

**Comment by chvamshi-xilinx** [comment]
Incorporated all the feedback except calling-out some of experimental IOCTLS. @sonals  and @mamin506. Please review again

**Comment by chvamshi-xilinx** [comment]
@larry9523 . Thanks for the feedback. Updated comments according your feedback

**Comment by sonals** [comment]
Already merged as part of https://github.com/Xilinx/XRT/pull/1994

**Pull request closed without merging**